Support For Arbitrary Number (1+) of Keychains

thunderbiscuit opened the pull request:
# Description

This PR allows the `Wallet` type to track any number of keychains. It is a breaking change.

Related Issues: #188, #227
Related PRs: #230, #226
Exploratory Crate: [multi-keychain-wallet](https://github.com/ValuedMammal/multi-keychain-wallet)

Follow along the todos and PR development on [this HackMD file](https://hackmd.io/@bdk/318).

<br>

# Commits

| Commit | Changes |
| ------ | ------ |
| 1 | Comment out `Wallet` dependent code. |
| 2 | Add `KeyRing` and related types. |
| 3 | Modify `Wallet` to levergage the `KeyRing`. |
| 4 | Add docs and examples. |
| 5 | Add tests. |

<br>

# Todos

Search for the following pattern to find Todos in the codebase:

```rust
// TODO PR #318: We should fix this because...
```

<br>

# Changelog notice

TODO

thunderbiscuit reviewed (commented):
Quick review on 6c50f002f1ffaaea58db25fc65bf5b189ec4896f: given your commit message I thought you were adding the Wallet::new constructor here, but the commit only contains the addition of the KeyRing to the Wallet type. Mind you this might be enough for this commit. Feel free to add a new commit with the constructor, or add it to this commit; whatever you think works best.

Also note that fa984e6c0ea6d59ff68bc5dc8d984bea8de190ea and 6c50f002f1ffaaea58db25fc65bf5b189ec4896f don't compile because of 

```rust
impl AsRef<bdk_chain::tx_graph::TxGraph<ConfirmationBlockTime>> for Wallet {
    fn as_ref(&self) -> &bdk_chain::tx_graph::TxGraph<ConfirmationBlockTime> {
        self.keychain_tx_graph.graph()
    }
}
```

on line 2638 of wallet/mod.rs. You could comment it out with the rest to ensure everything builds.

110CodingP commented on wallet/src/wallet/changeset.rs:
```diff
@@ -101,58 +103,47 @@ type IndexedTxGraphChangeSet =
+    pub keyring: keyring::changeset::ChangeSet<K>,
     /// Changes to the [`LocalChain`](local_chain::LocalChain).
-    pub local_chain: local_chain::ChangeSet,
+    pub chain: local_chain::ChangeSet,
```
I did the rename since the corresponding field in `Wallet` is so. reverted the rename in [588121b](https://github.com/bitcoindevkit/bdk_wallet/pull/318/commits/588121b2f269c8bf9a4a6269b7170644e786440f).

110CodingP commented:
I am so sorry :cry: . I completely messed dividing the commits into two. Things should be fixed now!

thunderbiscuit commented:
This one now needs a big ol' rebase @110CodingP.

110CodingP commented:
Wow! this one was HUGE! For the first few commits I did something like `git rebase --onto master HEAD~7 HEAD~6` and for the later ones I cherry-picked each commit onto master and finally did a `reset --hard`, wondering if there's a simpler way to do difficult rebases :sweat_smile: ?

thunderbiscuit commented:
Adding an example like this fails:

```rust
    // Simple KeyRing, allowing us to build a standard 2-descriptor wallet.
    let external_descriptor: &str = "tr(tpubD6NzVbkrYhZ4WyC5VZLuSJQ14uwfUbus7oAFurAFkZA5N3groeQqtW65m8pG1TT1arPpfWu9RbBsc5rSBncrX2d84BAwJJHQfaRjnMCQwuT/86h/1h/0h/0/*)";
    let internal_descriptor: &str = "tr(tpubD6NzVbkrYhZ4WyC5VZLuSJQ14uwfUbus7oAFurAFkZA5N3groeQqtW65m8pG1TT1arPpfWu9RbBsc5rSBncrX2d84BAwJJHQfaRjnMCQwuT/86h/1h/0h/1/*)";

    let mut keyring: KeyRing<KeychainKind> = KeyRing::new(Network::Regtest, KeychainKind::External, external_descriptor);
    keyring.add_descriptor(KeychainKind::Internal, internal_descriptor, false);

    let mut wallet = Wallet::new(keyring);
```

With a stacktrace containing:
```sh
thread 'main' panicked at /Users/user/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/miniscript-12.3.5/src/descriptor/key.rs:687:14:
The key should not contain any wildcards at this point
```

After a bit of poking around I found that this is coming from the descriptor_id() method called inside insert_descriptor():

```rust
    pub fn insert_descriptor(
        &mut self,
        keychain: K,
        descriptor: Descriptor<DescriptorPublicKey>,
    ) -> Result<bool, InsertDescriptorError<K>> {
        let did = descriptor.descriptor_id();
```

I don't have time to finish the investigation today, but just pointing it out. I'm not sure yet why this worked well in the multi-keychain-wallet crate and not here, nor why the code for the constructor is so different than the other previous constructors like `Wallet::create_with_params`.

thunderbiscuit commented:
Total facepalm. Sorry @110CodingP! Ok so I simplified the example a little bit just to keep a really neat version of how to build a wallet as close as you can from the 2.0 approach.

thunderbiscuit committed:
docs: add `KeyRing` based examples

Also fixed original examples.
Co-authored-by: codingp110 <codingp110@gmail.com>
Co-authored-by: valued mammal <valuedmammal@protonmail.com>

110CodingP commented:
Also is `balance_per_keychain` api desirable? ig it would help with this [bdk#57](https://github.com/bitcoindevkit/bdk/issues/57#issue-700190681).

luisschwab commented:
I'm thinking about how this will interact with the wallet's birthday, since different keychains can possibly have different birthdays (e.g. adding a keychain with older transactions to a wallet with a keychain that has newer transactions).

110CodingP commented:
> I noticed you have the sync examples commented out for now. When making synching a wallet, are you planning to add all revealed spk's into an "spk pool" and sync everything, or are we going to have to specify which keychain we want to sync? 

Currently we sync all keychains in [one go](https://github.com/thunderbiscuit/bdk_wallet/blob/25a41f33a01f2dd46c24fddeb40822db79effe41/src/wallet/mod.rs#L791) . But ig having an API to sync a particular keychain would also be great in case of large wallets with multiple keychains? 

> I ask this because with the first option, I can just take the first transaction from the "spk pool" and set the wallet's birthday as the transaction's `BlockId`, instead of possibly having to overwrite it if another keychain has an older birthday and we sync it later.

Yup, I agree this should work if we are not allowing addition of keychains!

thunderbiscuit committed:
tests: add tests for KeyRing and Wallet

Co-authored-by: codingp110 <codingp110@gmail.com>
Co-authored-by: valued mammal <valuedmammal@protonmail.com>

evanlinjin commented:
@110CodingP thanks for the reply. This comment is my understanding of the situation. That said, I shouldn't have the final say with bdk_wallet. I will say that I never got a "this approach is bad" response to my suggestion — but maybe there are flaws with my thinking here which aren't voiced. It'd be worth checking with the other maintainers.

110CodingP commented on src/wallet/mod.rs:
```diff
@@ -583,36 +615,58 @@ where
         self.indexed_graph.index.index_of_spk(spk).cloned()
     }
+
+    /// Get unbounded script pubkey iterators for both `Internal` and `External` keychains.
```
nit: `Internal` and `External` should be removed?

thunderbiscuit commented on src/wallet/persisted.rs:
```diff
@@ -380,7 +376,7 @@ pub enum LoadWithPersistError<E, K> {
     Persist(E),
     /// Occurs when the loaded changeset cannot construct [`Wallet`].
-    InvalidChangeSet(LoadError<K>),
+    InvalidChangeSet(crate::LoadError<K>),
```
Yep thanks!

110CodingP committed:
feat!: add `KeyRing` + related types + fns

Added `keyring::ChangeSet`.
The `add_descriptor` function returns a `KeyRing::Changeset` since
this would eventually be required when we introducing APIs to add
descriptors  to `Wallet.keyring`.
Also implemented `FromSql` and `ToSql` for `KeychainKind`.
Also modified `DescriptorError` and added `KeyRingError` since
information of whether a keychain/desc is already assigned should be
with the wallet and thus belongs to its error types.

Co-authored-by: thunderbiscuit <thunderbiscuit@protonmail.com>
Co-authored-by: valued mammal <valuedmammal@protonmail.com>

110CodingP committed:
refactor!: comment out `Wallet` dependent code

110CodingP committed:
feat!: modify Wallet struct to hold `KeyRing`

Modified the `Wallet::ChangeSet` accordingly.
Modified AddressInfo, add reveal_next_address
Also added default version for address APIs.
Modified Update to take in a generic
Also modified `balance` method to incorporate `CanonicalView`
and a more flexible `trust_predicate` which fixes incorrect
trusted balance calculation by the `Wallet`.
Modified `CreateParams` to incorporate the changes.
Modified `persist_test_utils` to incorporate the new `Wallet`. Note:
`persist_keychains` and `persist_keychain` are two separate tests
keeping in mind that some users may not use > 1 keychain.
Among all tests only `persist_keychains` assumes multiple keychains.
Co-authored-by: thunderbiscuit <thunderbiscuit@protonmail.com>
Co-authored-by: valued mammal <valuedmammal@protonmail.com>

thunderbiscuit reviewed (commented):
I found some methods that are cut out from the codebase because of commit 5df393d4e1329e9dd9e46c5dad099ed8a857797b, we should make sure we bring those back (or remove only the ones we really want to remove).

thunderbiscuit commented on src/wallet/mod.rs:
```diff
@@ -2572,217 +1437,1009 @@ impl Wallet {
-    ///
-    /// If `f` returns an error, then returns `E` of a type defined by the function
-    /// passed in.
-    pub fn events_helper<F, T, E>(&mut self, f: F) -> Result<Vec<WalletEvent>, E>
```
Some methods are removed entirely from the module, for example here `Wallet::events_helper`.

110CodingP replied:
Thanks for catching this! This shouldn't have happened though because this is just a rebase over master. Trying to figure out the reason.

thunderbiscuit committed:
temp: bring back all the events-related methods

thunderbiscuit committed:
temp: add Wallet::public_descriptor method
